subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java

```diff
@@ -262,6 +292,99 @@ private void populate(Project project) {
         }
     }
 
+    private Map<String, String> gatherClosedProjectPaths(Set<String> closedProjectNames) {
+        Map<String, String> closedProjectPaths = new HashMap<>();
+        for (ProjectState state : projectRegistry.getAllProjects()) {
+            ProjectInternal project = state.getOwner().getLoadedSettings().getGradle().getRootProject().project(state.getComponentIdentifier().getProjectPath());
+            EclipseModel model = project.getExtensions().getByType(EclipseModel.class);
+            if (closedProjectNames.contains(model.getProject().getName())) {
+                closedProjectPaths.put(model.getProject().getName(), project.getIdentityPath().getPath());
+            }
+        }
+        return closedProjectPaths;
+
+    }
+
+    private List<DefaultEclipseExternalDependency> createSubstituteDependencies(Project project, ProjectDependency projectDependency) {
```

Comment:
❌ If we need additional information from the dependency resolver, like what artifacts a project dependency can be replaced by, we should add it [there](https://github.com/gradle/gradle/blob/master/subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/EclipseDependenciesCreator.java#L100), not reimplement it here.

Reply:
so, something along the lines of what I tried in my previous attempt? 
(see https://github.com/gradle/gradle/compare/master...fraenkelc:project_dependency_substitution?diff=unified)
What I did there was:
* extend ProjectDependency with attributes for the publication(/file) and the build dependencies (+ I had to fix the equals as otherwise I would not get multiple dependencies for different artifacts from the same project for differnt configurations)
* extract that information  from the `ComponentArtifactMetadata` in `EclipseDependenciesCreator` and pass it along to the `ProjectDependencyBuilder`
* use that information in `EclipseModelBuilder`

Reply:
Yes, that looks a lot better 👍 

Just curious, you said this was your previous attempt: What made you drop that attempt? I didn't follow the whole history of this PR, only the final version.

Reply:
There was no technical reason(I haven't tried composites with that branch yet though) , probably just a misunderstanding on my part. I thought https://github.com/eclipse/buildship/issues/881 was asking for a solution that's local to EclipseModelBuilder and that does it's own artifact views